File: ecolink-front/src/app/core/services/complaint.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Complaint, ComplaintPriority, ComplaintStatus } from '../models/index';
import { environment } from '../../../environments/environment';

@Injectable({ providedIn: 'root' })
export class ComplaintService {
  private base = `${environment.apiUrl}/complaints`;
  constructor(private http: HttpClient) {}

  create(complaint: Partial<Complaint>): Observable<Complaint> {
    return this.http.post<Complaint>(this.base, complaint);
  }
  getById(id: number): Observable<Complaint> {
    return this.http.get<Complaint>(`${this.base}/${id}`);
  }
  getByUserId(userId: number): Observable<Complaint[]> {
    return this.http.get<Complaint[]>(`${this.base}/user/${userId}`);
  }
  getByStatus(status: ComplaintStatus): Observable<Complaint[]> {
    return this.http.get<Complaint[]>(`${this.base}/status/${status}`);
  }
  getByPriority(priority: ComplaintPriority): Observable<Complaint[]> {
    return this.http.get<Complaint[]>(`${this.base}/priority/${priority}`);
  }
  update(id: number, complaint: Partial<Complaint>): Observable<Complaint> {
    return this.http.put<Complaint>(`${this.base}/${id}`, complaint);
  }
  updateStatus(id: number, status: ComplaintStatus): Observable<Complaint> {
    const params = new HttpParams().set('status', status);
    return this.http.put<Complaint>(`${this.base}/${id}/status`, null, { params });
  }
}

File: flutter_application_1/lib/services/complaint_service.dart
import '../models/complaint.dart';
import 'api_client.dart';

class ComplaintService {
  static Future<List<Complaint>> getByUserId(int userId) async {
    final json = await ApiClient.get('/complaints/user/$userId');
    return (json as List? ?? [])
        .map((e) => Complaint.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  static Future<Complaint> create(
      Complaint complaint, Map<String, dynamic> userRef) async {
    final json =
        await ApiClient.post('/complaints', body: complaint.toCreateJson(userRef));
    return Complaint.fromJson(json as Map<String, dynamic>);
  }
}
